### PR TITLE
Do not allow Lenovo hardware to install multiple capsules

### PR DESCRIPTION
**Pull request by hughsie**
Type of pull request:
- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [X] Feature
- [ ] Documentation

### CODE REVIEW

**Review by superm1 — APPROVED** [review]
The PR looks good for the Lenovo case.  I think you want to set `FWUPD_UPDATE_STATE_PENDING` for the stuff pending activation too though in the engine.

**Comment by hughsie** [comment]
Something like this?

    -       if (fu_device_get_update_state (device) == FWUPD_UPDATE_STATE_PENDING) {
    +       if (fu_device_get_update_state (device) == FWUPD_UPDATE_STATE_PENDING ||
    +           fu_device_has_flag (device, FWUPD_DEVICE_FLAG_NEEDS_ACTIVATION)) {

**Comment by superm1** [comment]
I was thinking directly in the plugins rather than the engine initially, but yeah I suppose that will work.

**Comment by superm1** [comment]
Something else I realize needed now - if activate is called, those flags need to be cleared still.

**Comment by hughsie** [comment]
> if activate is called, those flags need to be cleared still

Not so convinced, and it's a bit of long journey explaining why...

Rather than setting the plugin flag in `fu_engine_update_cleanup()` you could do:

    +       g_signal_connect (device, "notify::flags",
    +                         G_CALLBACK (fu_engine_flags_notify_cb), self);
    +       g_signal_connect (device, "notify::update-state",
    +                         G_CALLBACK (fu_engine_flags_notify_cb), self);

...and then check for the `FWUPD_UPDATE_STATE_PENDING` state or `FWUPD_DEVICE_FLAG_NEEDS_ACTIVATION` flags there. That's a bit less magic than just at the end of update function and is probably a good move.

The next logical jump is to say "we could clear the flag if those device flags are not set" but you of course need to query all the active FuDevice objects for that specific plugin (e.g. `fu_device_list_get_active()`, then filter by plugin) and then clear the flag if none of them require activation or are pending. That would work.

But! Clearing the flag there undoes the work when we've restarted fwupd after deploying a UEFI update -- the plugin sets the flag when `FU_EFIVAR_GUID_EFI_GLOBAL:BootNext` and `FU_EFIVAR_GUID_FWUPDATE-*` exist -- none of the devices are set to `FWUPD_UPDATE_STATE_PENDING` as you might expect.

So, we can either refcount the flag as two things are setting it (ewww!) or we can untangle the mess a little. Or we can just avoid unsetting the flag on activation; which seems to be a _nice to have_ thing as activation is typically done just before we shutdown and then everything is killed anyway. Or, we go back a bit and define `FWUPD_PLUGIN_FLAG_PENDING_UPDATE` as nothing to do with activation and then this isn't a problem again. Other ideas welcome.

File: libfwupd/fwupd-device.c [hunk 1]
```diff
@@ -74,6 +74,7 @@ enum {
 	PROP_PROTOCOL,
 	PROP_STATUS,
 	PROP_PARENT,
+	PROP_UPDATE_STATE,
```
Comment (superm1) [inline libfwupd/fwupd-device.c:77]: With this type of change I almost need to wonder if we really need to be using the device flag for needs activation.  Should we deprecate it and become an update state?

File: src/fu-engine.c [hunk 2]
```diff
@@ -2696,6 +2715,9 @@ fu_engine_activate (FuEngine *self, const gchar *device_id, GError **error)
 	if (!fu_plugin_runner_activate (plugin, device, error))
 		return FALSE;
 
+	/* unconditionally do this on success */
+	fu_device_remove_flag (device, FWUPD_DEVICE_FLAG_NEEDS_ACTIVATION);
```
Comment (superm1) [inline src/fu-engine.c:2719]: I don't think this is correct, this is already done in `fu_plugin_runner_activate`

**Comment by superm1** [comment]
>  typically done just before we shutdown and then everything is killed anyway

**Typically** is _relative_. It's also done at unplug for Dell's WD19 as well and logout for ChromeOS.

File: src/fu-engine.c (hunk 2 above)
Comment (hughsie) [reply src/fu-engine.c:2719]: Agree, dropped, thanks.

**Comment by hughsie** [comment]
> do we even need a "public" daemon plugin flag in the first place

I think this is the right question, and I think the answer is no. New PRs coming up, thanks for the sanity check yet again.

**Comment by hughsie** [comment]
I think https://github.com/fwupd/fwupd/pull/2757 is a much better solution.

**Pull request closed without merging**